### PR TITLE
Add api for the beginning and the end of presence

### PR DESCRIPTION
**Pull request by marcindbar**
[Ticket](https://kindergarten.stxnext.pl/browse/WDP-81)
[Pylint](https://gist.github.com/marcindbar/081ee0102f7865ec8ca791adfd67f0c8)

### CODE REVIEW

File: src/presence_analyzer/static/presence_start_end.html [hunk 1]
```diff
@@ -79,6 +79,13 @@
         google.load("visualization", "1", {packages:["corechart", "timeline"], 'language': 'pl'});
     </script>
     <script type="text/javascript">
+        function parseInterval(value) {
+            var result = new Date(1,1,1);
+            result.setMilliseconds(value*1000);
+            console.log(result);
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]: ᗒ ͟ʖᗕ

File: src/presence_analyzer/static/presence_start_end.html [hunk 2]
```diff
@@ -79,6 +79,13 @@
         google.load("visualization", "1", {packages:["corechart", "timeline"], 'language': 'pl'});
     </script>
     <script type="text/javascript">
+        function parseInterval(value) {
+            var result = new Date(1,1,1);
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]: spaces after comma

File: src/presence_analyzer/static/presence_start_end.html [hunk 3]
```diff
@@ -96,33 +103,31 @@
...
-                        };
-                        var formatter = new google.visualization.DateFormat({pattern: 'HH:mm:ss'});
-                        formatter.format(data, 1);
-                        formatter.format(data, 2);
+                        $.getJSON("/api/v1/presence_start_end/"+selected_user, function(result) {
+                            $.each(result, function(index, value) {
+                                value[1] = parseInterval(value[1]);
+                                value[2] = parseInterval(value[2]);
+                            });
+                            console.log(result);
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]: ᗒ ͟ʖᗕ

File: src/presence_analyzer/static/presence_start_end.html [hunk 4]
```diff
@@ -96,33 +103,31 @@
...
+                            var options = {
+                                hAxis: {title: 'Weekday'}
+                            };
+                            var formatter = new google.visualization.DateFormat({pattern: 'HH:mm:ss'});
+                            formatter.format(data, 1);
+                            formatter.format(data, 2);
 
+                            chart_div.show();
+                            loading.hide();
+                            var chart = new google.visualization.Timeline(chart_div[0]);
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]: single var for all variables + declare at the top of the scope what u can (after $.getJSON)

File: src/presence_analyzer/views.py [hunk 5]
```diff
@@ -8,7 +8,8 @@
 from flask import redirect, abort
 
 from presence_analyzer.main import app
-from presence_analyzer.utils import jsonify, get_data, mean, group_by_weekday
+from presence_analyzer.utils import jsonify, get_data, mean, group_by_weekday,\
+    group_by_weekday_start_end
```
Comment (piotrgrundas) [inline src/presence_analyzer/views.py]: from presence_analyzer.utils import (
        jsonify,
        ...
    )

File: src/presence_analyzer/static/presence_start_end.html [hunk 6]
```diff
@@ -79,6 +79,13 @@
         google.load("visualization", "1", {packages:["corechart", "timeline"], 'language': 'pl'});
     </script>
     <script type="text/javascript">
+        function parseInterval(value) {
+            var result = new Date(1,1,1);
+            result.setMilliseconds(value*1000);
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]: ditto

File: src/presence_analyzer/views.py [hunk 7]
```diff
@@ -73,3 +75,22 @@ def presence_weekday_view(user_id):
 
     result.insert(0, ('Weekday', 'Presence (s)'))
     return result
+
+
+@app.route('/api/v1/presence_start_end/<int:user_id>', methods=['GET'])
+@jsonify
+def mean_time_presence_weekday_view(user_id):
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: mean_time?

File: src/presence_analyzer/utils.py [hunk 8]
```diff
@@ -101,3 +104,16 @@ def mean(items):
     Calculates arithmetic mean. Returns zero for empty lists.
     """
     return float(sum(items)) / len(items) if len(items) > 0 else 0
+
+
+def group_by_weekday_start_end(items, info):
```
Comment (dpolcyn) [inline src/presence_analyzer/utils.py]: info? I dont understand what it is

File: src/presence_analyzer/tests.py [hunk 9]
```diff
@@ -61,13 +61,13 @@ def test_mean_time_weekday_view(self):
         Test mean time weekday view.
         """
         weekdays_means = [
-            ['Mon', 19733.666666666668],
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: I don't see any reason of this changes

File: src/presence_analyzer/static/mean_time_weekday.html [hunk 10]
```diff
@@ -23,7 +23,7 @@
             padding: 0 2em;
             border: 1px solid #AAA;
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/mean_time_weekday.html]: lowercase

File: src/presence_analyzer/static/mean_time_weekday.html [hunk 11]
```diff
@@ -80,47 +80,50 @@
...
+                                data
+                                formatter,
+                                options;
+
                             $.each(result, function(index, value) {
                                 value[1] = parseInterval(value[1]);
                             });
-                            console.log(result);
-                            var data = new google.visualization.DataTable();
+                            data = new google.visualization.DataTable();
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/mean_time_weekday.html]: U can assign it directly in 107

File: src/presence_analyzer/static/mean_time_weekday.html [hunk 12]
```diff
@@ -80,47 +80,50 @@
...
                             data.addColumn('string', 'Weekday');
                             data.addColumn('datetime', 'Mean time (h:m:s)');
                             data.addRows(result);
-                            var options = {
+                            options = {
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/mean_time_weekday.html]: ditto

File: src/presence_analyzer/static/mean_time_weekday.html [hunk 13]
```diff
@@ -80,47 +80,50 @@
...
                                 hAxis: {title: 'Weekday'}
                             };
-                            var formatter = new google.visualization.DateFormat({pattern: 'HH:mm:ss'});
+                            formatter = new google.visualization.DateFormat({pattern: 'HH:mm:ss'});
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/mean_time_weekday.html]: ditto

File: src/presence_analyzer/static/mean_time_weekday.html [hunk 14]
```diff
@@ -80,47 +80,50 @@
...
                             formatter.format(data, 1);
 
-                            
                             chart_div.show();
                             loading.hide();
-                            var chart = new google.visualization.ColumnChart(chart_div[0]);
+                            chart = new google.visualization.ColumnChart(chart_div[0]);
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/mean_time_weekday.html]: again

File: src/presence_analyzer/static/mean_time_weekday.html [hunk 15]
```diff
@@ -80,47 +80,50 @@
...
             return result;
         }
 
         (function($) {
             $(document).ready(function(){
                 var loading = $('#loading');
-                $.getJSON("/api/v1/users", function(result) {
-                    var dropdown = $("#user_id");
+                $.getJSON('/api/v1/users', function(result) {
+                    var dropdown = $('#user_id');
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/mean_time_weekday.html]: If u assign jquery selector to variable proceed it with $ - `$dropdown`
Also in html don't use underscore - use dash instead

File: src/presence_analyzer/static/presence_start_end.html [hunk 16]
```diff
@@ -79,50 +79,58 @@
...
         (function($) {
             $(document).ready(function(){
                 var loading = $('#loading');
-                $.getJSON("/api/v1/users", function(result) {
-                    var dropdown = $("#user_id");
+                $.getJSON('/api/v1/users', function(result) {
+                    var dropdown = $('#user_id');
                     $.each(result, function(item) {
-                        dropdown.append($("<option />").val(this.user_id).text(this.name));
+                        dropdown.append($('<option/>').val(this.user_id).text(this.name));
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/presence_start_end.html]:
```
$dropdown.append($('<option/>', {
    'val': this.user_id,
    'text': this.name
});
```

File: src/presence_analyzer/utils.py [hunk 17]
```diff
@@ -101,3 +104,20 @@ def mean(items):
...
     return float(sum(items)) / len(items) if len(items) > 0 else 0
+
+
+def group_by_weekday_start_end(items):
+    """
+    Groups the beginnings of the ends of presence entries by weekday.
+    """
+    result_start, result_end = 7 * [[]], 7 * [[]]
+    for date in items:
+        seconds_start = seconds_since_midnight(items[date]['start'])
```
Comment (dpolcyn) [inline src/presence_analyzer/utils.py]: Please rewrite this function. Use dicts. This is mess

File: src/presence_analyzer/views.py [hunk 18]
```diff
@@ -73,3 +75,39 @@ def presence_weekday_view(user_id):
 
     result.insert(0, ('Weekday', 'Presence (s)'))
     return result
+
+
+# @app.route('/api/v1/presence_start_end/<int:user_id>', methods=['GET'])
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: ?????????????????????????????
Why you commit commented code?

File: src/presence_analyzer/utils.py [hunk 19]
```diff
@@ -101,3 +105,25 @@ def mean(items):
     Calculates arithmetic mean. Returns zero for empty lists.
     """
     return float(sum(items)) / len(items) if len(items) > 0 else 0
+
+
+def group_by_weekday_start_end(items):
+    """
+    Groups the beginnings of the ends of presence entries by weekday.
+    """
+    result = {nr: {'start': [], 'end': []} for nr in range(7)}
```
Comment (dpolcyn) [inline src/presence_analyzer/utils.py]: i instead of nr

File: src/presence_analyzer/views.py [hunk 20]
```diff
@@ -8,7 +8,10 @@
 from flask import redirect, abort
 
 from presence_analyzer.main import app
-from presence_analyzer.utils import jsonify, get_data, mean, group_by_weekday
+from presence_analyzer.utils import (
+    jsonify, get_data, mean, group_by_weekday, group_by_weekday_start_end,
```
Comment (dpolcyn) [inline src/presence_analyzer/views.py]: sort imports

File: src/presence_analyzer/tests.py [hunk 21]
```diff
@@ -113,6 +113,26 @@ def test_presence_weekday_view_user_do_not_exist(self):
 
         self.assertEqual(resp.status_code, 404)
 
+    def test_group_by_weekday_start_end(self):
+        """
+        Test getting interval of mean presence time grouped by weekday
+        """
+        weekdays_means = [
+            ["Mon", 33134.0, 57257.0],
```
Comment (dpolcyn) [inline src/presence_analyzer/tests.py]: we use ' in python

File: src/presence_analyzer/static/mean_time_weekday.html [hunk 22]
```diff
@@ -80,47 +80,49 @@
...
-                $('#user_id').change(function(){
-                    var selected_user = $("#user_id").val();
-                    var chart_div = $('#chart_div');
-                    if(selected_user) {
-                        loading.show();
-                        chart_div.hide();
-                        $.getJSON("/api/v1/mean_time_weekday/"+selected_user, function(result) {
+                $('#user-id').change(function(){
+                    var selectedUser = $('#user-id').val(),
+                        $chartDiv = $('#chart-div');
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/mean_time_weekday.html]: enter for readability.

File: src/presence_analyzer/static/mean_time_weekday.html [hunk 23]
```diff
@@ -80,47 +80,49 @@
...
+                    if(selectedUser) {
+                        $loading.show();
+                        $chartDiv.hide();
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/mean_time_weekday.html]: ditto

File: src/presence_analyzer/static/mean_time_weekday.html [hunk 24]
```diff
@@ -80,47 +80,49 @@
...
-                            var formatter = new google.visualization.DateFormat({pattern: 'HH:mm:ss'});
                             formatter.format(data, 1);
 
-                            
-                            chart_div.show();
-                            loading.hide();
-                            var chart = new google.visualization.ColumnChart(chart_div[0]);
+                            $chartDiv.show();
+                            $loading.hide();
+                            chart = new google.visualization.ColumnChart($chartDiv[0]);
```
Comment (piotrgrundas) [inline src/presence_analyzer/static/mean_time_weekday.html]: I think u can assign this straight upon declaration.

File: src/presence_analyzer/utils.py [hunk 25]
```diff
@@ -4,13 +4,15 @@
...
 import logging
+from copy import copy
 from datetime import datetime
 from functools import wraps
 from json import dumps
 
 from flask import Response
 
 from presence_analyzer.main import app
+from calendar import day_abbr
```
Comment (piotrgrundas) [inline src/presence_analyzer/utils.py]: sort imports

File: src/presence_analyzer/views.py [hunk 26]
```diff
@@ -8,7 +8,10 @@
 from flask import redirect, abort
 
 from presence_analyzer.main import app
-from presence_analyzer.utils import jsonify, get_data, mean, group_by_weekday
+from presence_analyzer.utils import (
+    get_data, group_by_weekday, group_by_weekday_start_end, jsonify, mean,
+    mean_by_weekday
+)
```
Comment (piotrgrundas) [inline src/presence_analyzer/views.py]: You were told before how to deal with longer list of imports - this is wrong.

File: src/presence_analyzer/views.py [hunk 27]
```diff
@@ -73,3 +76,19 @@ def presence_weekday_view(user_id):
...
+    Returns interval of mean presence time of given user grouped by weekday.
+    """
+    data = get_data()
+    if user_id not in data:
+        log.debug('User %s not found!', user_id)
+        abort(404)
+
+    weekdays_data = group_by_weekday_start_end(data[user_id])
+    result = [mean_by_weekday(key, val) for key, val in weekdays_data.items()]
+    return result
```
Comment (piotrgrundas) [inline src/presence_analyzer/views.py]: Personally I would return directly list comprehension  instead assign it to variable just to then return it.

**Pull request merged**